ast/parser_ext.go

```diff
@@ -642,6 +642,15 @@ func parseModule(filename string, stmts []Statement, comments []*Comment) (*Modu
 				continue
 			}
 			mod.Rules = append(mod.Rules, rule)
+			// NOTE(trolloldem): a Rule created through ParseRuleFromBody
+			// does not have the correct JSONOptions setted
+			vis := NewGenericVisitor(func(x interface{}) bool {
+				if x, ok := x.(customJSON); ok {
+					x.setJSONOptions(stmt[0].jsonOptions)
```

Comment:
This would be same as `rule.jsonOptions`?

Reply:
This is what the PR is correcting: in this point of the code the `rule` variable is a newly created `Rule` struct. In the case reported in #5790, the struct is created in [ParseCompleteDocRuleFromEqExpr](https://github.com/open-policy-agent/opa/blob/main/ast/parser_ext.go#L243). The returned `rule` has not the field `jsonOptions` correctly set, differently from the value of `stmt[0]`. `smts` in this branch of the switch case has [Body](https://github.com/open-policy-agent/opa/blob/main/ast/policy.go#L214) type and `smts[0]` is an [Expr](https://github.com/open-policy-agent/opa/blob/main/ast/policy.go#L217) struct with the correct field `jsonOptions`.

The `visitor` is used to correctly set `jsonOptions` also in structs that are fields of `rule` (e.g. its [Head](https://github.com/open-policy-agent/opa/blob/main/ast/policy.go#L182))

Reply:
So the body in this case has only one expression which is interpreted as a rule. 

> The returned rule has not the field jsonOptions correctly set

I'm curious why we add this change in this method instead of ParseRule* methods. It would be helpful if you could clarify. Thanks.

Reply:
I think it is possible to move this change into the `ParseRuleFromBody` function if preferred, since the logic used with `stmt[0]` can be applied to `body[0]` without any change to existing interfaces. For the moment I've chosen to set this operation in the switch statement to highlight the necessity to correctly set the `jsonOptions` field, but I like the proposal of returning an already correctly configured `Rule` struct

Reply:
One could argue that for any public function that parses a rule from something that carries `JSONOptions`, then those should be reflected in the returned struct. Even though the implementation won't be as nicely contained as it is here, I think I agree with @ashutosh-narkar : we should update all the `Parse*From*` functions to transfer `JSONOptions`. That way, we keep the API consistent.

Reply:
This seems good to me. I will try to set the correct value of `jsonOptions` in the various `Parse*From*` functions (e.g. [ParseCompleteDocRuleFromEqExpr](https://github.com/open-policy-agent/opa/blob/main/ast/parser_ext.go#L243))

Reply:
As discussed, I have moved the steps necessary to have a Rule with the correct value `jsonOptions` set directly in the publicly exposed API. To do this I have added a not exported function `setTermsJSONOptions`, to handle the procedure needed to set all the `JSONOptions`  fields in the various substructs. 

I have also added a new test `TestRuleFromBodyJSONOptions` and a corresponding assertion `assertParseModuleJSONOptions`. This test uses the same cases used to check the correctness of the parsing procedure in `TestRuleFromBody`. The new assertion is added since the `jsonOptions` field is not used to establish if two `Rule` structs are equal, thus this is explicitly checked in the new assertion. I have added a test case not present in `TestRuleFromBody` (i.e. `foo.fizz.buzz`) to test the execution path using `ParseCompleteDocRuleWithDotsFromTerm`.